mesh/mesh.hpp

```diff
@@ -1014,6 +1022,138 @@ class Mesh
+       @note Duplicate entries will be ignored and the resulting sets will be
+       sorted.
+    */
+   void AddToAttributeSet(const std::string &set_name,
+                          const Array<int> &attr);
+
+   /// @brief Add a single entry to a boundary attribute set
+   /**
+       @param[in] set_name The name of the set being augmented
+       @param[in] attr A single attribute number to be inserted in the set
+
+       @note If the named set does not exist a new set will be created.
+       @note Duplicate entries will be ignored and the resulting sets will be
+       sorted.
+    */
+   void AddToBdrAttributeSet(const std::string &set_name, int attr);
+
+   /// @brief Add an array of entries to a boundary attribute set
+   /**
+       @param[in] set_name The name of the set being augmented
+       @param[in] attr Array of attribute numbers to be inserted in the set
+
+       @note If the named set does not exist a new set will be created.
+       @note Duplicate entries will be ignored and the resulting sets will be
+       sorted.
+    */
+   void AddToBdrAttributeSet(const std::string &set_name,
+                             const Array<int> &attr);
+
+   /// @brief Remove a single entry from an attribute set
+   /**
+       @param[in] set_name The name of the set being modified
+       @param[in] attr A single attribute number to be removed from the set
+
+       @note If the named set does not exist a new empty set will be created.
+       @note If @a attr is not a member of the named set the set will not
+       be modified and no error will occur.
+    */
+   void RemoveFromAttributeSet(const std::string &set_name, int attr);
+
+   /// @brief Remove a single entry from a boundary attribute set
+   /**
+       @param[in] set_name The name of the set being modified
+       @param[in] attr A single attribute number to be removed from the set
+
+       @note If the named set does not exist a new empty set will be created.
+       @note If @a attr is not a member of the named set the set will not
+       be modified and no error will occur.
+    */
+   void RemoveFromBdrAttributeSet(const std::string &set_name, int attr);
+
+   /// @brief Access a named attribute set
+   /**
+       @param[in] set_name The name of the set being accessed
+
+       @note If the named set does not exist a new empty set will be created
+       and returned.
+
+       @note The reference returned by this method can be invalidated by
+       subsequent calls to SetAttributeSet, ClearAttributeSet, or
+       RemoveFromAttributeSet. AddToAttributeSet should not invalidate this
+       reference.
+    */
+   Array<int> & GetAttributeSet(const std::string & set_name);
+
+   /// @brief Access a named boundary attribute set
+   /**
+       @param[in] set_name The name of the set being accessed
+
+       @note If the named set does not exist a new empty set will be created
+       and returned.
```

Comment:
minor/optional:

just wondering if there is a use case for this ? Alternatively a more defensive approach
```
Array<int> & Mesh::GetBdrAttributeSet(const std::string & set_name)
{
   return bdr_attr_sets.at(set_name);
}
```
could be used to avoid this side-effect (in which case the code will throw an exception).

Reply:
Probably not. I was thinking of justifying it as a way to create a new set when you don't yet know which attributes will belong to it. This is a flimsy justification.

To be honest I was avoiding the need to figure out how to handle the returned reference. I didn't want to halt execution in this situation. I considered issuing a warning message. But in this case what Array<int> do I return? I don't have an empty array in the Mesh class that will persist beyond the call to this method. I could add one as member data but that seems odd. I could allocate an array object just to return a reference to it but then the caller will need to delete it at some point. Do you (@paul-hilscher) happen to know how this situation is typically handled?

Reply:
I also think automatically creating a new empty attribute set is not the right behavior here. There are a couple of options to handle this:

* `MFEM_ABORT`. This will halt execution, but it probably indicates a bug that needs to be fixed anyway?
* Change the return type to `Array<int>*`, allowing for `nullptr` as a return value.
* When we eventually (hopefully soon!) move to C++17, we can start using `std::optional` (in this case it's a little more complicated since we would need to use `std::reference_wrapper` as well)

Reply:
In some of the simulation codes I'm familiar with, an attribute is checked for existence to determine if a particular kind of setup or physics is enabled, like slide lines. Having it return an empty set breaks the semantics, though it's not the end of the word.